tests/unit/source_code/test_s3fs.py

```diff
@@ -0,0 +1,152 @@
+from unittest.mock import create_autospec
+
+import pytest
+
+from databricks.labs.ucx.source_code.base import Advice, Deprecation
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.service.workspace import ObjectInfo, Language, ObjectType
+
+from databricks.labs.ucx.source_code.dependencies import DependencyLoader, SourceContainer, DependencyResolver
+from databricks.labs.ucx.source_code.notebook_migrator import NotebookMigrator
+from databricks.labs.ucx.source_code.whitelist import Whitelist
+from tests.unit import _load_sources, _download_side_effect
+
+S3FS_DEPRECATION_MESSAGE = "Use of dependency s3fs is deprecated"
+
+
+@pytest.mark.parametrize(
+    "source, expected",
+    [
+        (
+            "import s3fs",
+            [
+                Deprecation(
+                    code='dependency-check',
+                    message=S3FS_DEPRECATION_MESSAGE,
+                    start_line=0,
+                    start_col=0,
+                    end_line=0,
+                    end_col=0,
+                )
+            ],
+        ),
+        (
+            "from s3fs import something",
+            [
+                Deprecation(
+                    code='dependency-check',
+                    message=S3FS_DEPRECATION_MESSAGE,
+                    start_line=0,
+                    start_col=0,
+                    end_line=0,
+                    end_col=0,
+                )
+            ],
+        ),
+        ("import leeds", []),
+        ("from leeds import path", []),
+        (
+            "import s3fs, leeds",
+            [
+                Deprecation(
+                    code='dependency-check',
+                    message=S3FS_DEPRECATION_MESSAGE,
+                    start_line=0,
+                    start_col=0,
+                    end_line=0,
+                    end_col=0,
+                )
+            ],
+        ),
+        ("from leeds import path, s3fs", []),
+        (
+            "def func():\n    import s3fs",
+            [
+                Deprecation(
+                    code='dependency-check',
+                    message=S3FS_DEPRECATION_MESSAGE,
+                    start_line=0,
```

Comment:
No start line information is bad :(

Reply:
Yes. Hence two phases in implementation. The graph/tree needs additional information. Possibly best to add parent to dependency but that's why I probably want to change to a pass after dependency graph is built. Tokens are not current preserved, but will be.

Reply:
@nfx @jimidle I agree that it would be much simpler to proceed in at least 2 passes: a) build the dependency graph b) run the linters (might even need an intermediate one for dependency resolution)

Reply:
@ericvergnaud Yes - that is what I intend. It all kind of happens at once at the moment, and as we said, I will need to make changes for tracking so we know how we end up at a particular node in the graph.